moveit_planners/ompl/ompl_interface/src/detail/ompl_constraints.cpp

```diff
@@ -323,6 +323,134 @@ Eigen::MatrixXd OrientationConstraint::calcErrorJacobian(const Eigen::Ref<const
   return -angularVelocityToAngleAxis(aa.angle(), aa.axis()) * robotGeometricJacobian(x).bottomRows(3);
 }
 
+/******************************************
+ * Toolpath Constraint
+ * ****************************************/
+ToolPathConstraint::ToolPathConstraint(const moveit::core::RobotModelConstPtr& robot_model, const std::string& group,
+                                       unsigned int num_dofs, const unsigned int num_cons)
+  : BaseConstraint(robot_model, group, num_dofs, num_cons)
+{
+  pose_nn_.setDistanceFunction([this](const EigenIsometry3dWrapper& pose1, const EigenIsometry3dWrapper& pose2) {
+    const double translation_distance = (pose1.tform_.translation() - pose2.tform_.translation()).norm();
+    const double rotation_distance = arcLength(pose1.tform_, pose2.tform_);
+    return translation_distance + rotation_distance;
+  });
+}
+
+void ToolPathConstraint::parseConstraintMsg(const moveit_msgs::msg::Constraints& constraints)
+{
+  const auto path_constraint = constraints.path_constraints.at(0);
+  link_name_ = path_constraint.link_name;
+  bounds_ = pathConstraintMsgToBoundVector(path_constraint);
+  setPath(path_constraint);
+}
+
+void ToolPathConstraint::setPath(const moveit_msgs::msg::PathConstraint& path_constraint)
+{
+  // Clear previous path
+  path_.clear();
+  pose_nn_.clear();
+
+  // Create the new path
+  const auto num_pts = path_constraint.path.size();
+  if (num_pts > 1)
+  {
+    path_.reserve(num_pts);
+    for (const auto& pose : path_constraint.path)
+    {
+      Eigen::Isometry3d pose_eigen;
+      tf2::fromMsg(pose, pose_eigen);
+      path_.emplace_back(EigenIsometry3dWrapper(pose_eigen));
+    }
+  }
+  else
+  {
+    RCLCPP_ERROR_STREAM(LOGGER, "Path constraint must specify at least 2 points!");
+    return;
+  }
+
+  // Extract the interpolation distance
+  if (path_constraint.interp_distance > 0.0)
+  {
+    interp_distance_ = path_constraint.interp_distance;
+  }
+  else
+  {
+    RCLCPP_WARN_STREAM(LOGGER, "Cannot use zero interpolation distance. Defaulting to 0.01.");
+    interp_distance_ = 0.01;
```

Comment:
Given that you are discretizing the polyline at some resolution `interp_distance_`, the allowable constraint error has to be greater than this interpolation distance so that it would appear as a continuous function rather than a disjoint collection of points. Where is the constraint tolerance set in this example?

Reply:
I'm not sure what this is. Do you know off the top of your head what the OMPL name / function call would be so I can search for it?

Reply:
In [`ompl::base::Constraint`](https://ompl.kavrakilab.org/classompl_1_1base_1_1Constraint.html) this is the `setTolerance()` function - I'm not sure which solver MoveIt defaults to, but it's whatever is considered the cutoff for saying the constraint is satisfied.

Reply:
Aha, thanks! Yeah, it's just using the default of `1e-4` at the moment, so I can try bring that up.